File: android/src/main/java/com/margelo/nitro/wakeword/HybridWakeWord.kt
package com.margelo.nitro.wakeword

import android.Manifest
import android.content.pm.PackageManager
import android.util.Log
import androidx.annotation.Keep
import androidx.core.content.ContextCompat
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.modules.core.PermissionAwareActivity
import com.facebook.react.modules.core.PermissionListener
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise

@DoNotStrip
@Keep
class HybridWakeWord : HybridWakeWordSpec() {
  companion object {
    private const val PERMISSION_REQUEST_CODE = 0x5742
    private const val LOG_TAG = "NitroWakeWord"
  }

  private val context
    get() = NitroModules.applicationContext ?: throw WakeWordException("React context unavailable")

  @Volatile private var engine: OpenWakeWordEngine? = null
  private var capture: AudioCapture? = null
  private var foregroundService = false
  private var notificationTitle = "Listening"
  private var notificationText = "Wake word detection is active"
  private val lock = Any()

  @Volatile private var detectionListener: ((WakeWordDetection) -> Unit)? = null
  @Volatile private var scoreListener: ((String, Double) -> Unit)? = null
  @Volatile private var errorListener: ((String) -> Unit)? = null

  // Properties

  override val isLoaded: Boolean
    get() = engine != null

  override val isListening: Boolean
    get() = capture?.isRunning == true

  // Lifecycle

  override fun load(config: WakeWordConfig): Promise<Unit> = Promise.async {
    stop()
    if (config.models.isEmpty()) throw WakeWordException("config.models must contain at least one model")
    val ctx = context
    val vadThreshold = (config.vadThreshold ?: 0.3).toFloat()
    val specs = config.models.map { model ->
      OpenWakeWordEngine.KeywordSpec(
        keyword = model.keyword ?: ModelResolver.defaultKeyword(model.model),
        bytes = ModelResolver.load(ctx, model.model),
        threshold = (model.threshold ?: 0.5).toFloat(),
        patience = (model.patience ?: 1.0).toInt(),
      )
    }
    val built = OpenWakeWordEngine(
      melBytes = ModelResolver.load(ctx, "melspectrogram.onnx"),
      embeddingBytes = ModelResolver.load(ctx, "embedding_model.onnx"),
      vadBytes = if (vadThreshold > 0f) ModelResolver.load(ctx, "silero_vad.onnx") else null,
      vadThreshold = vadThreshold,
      refractoryMs = config.refractoryMs ?: 1000.0,
      keywordSpecs = specs,
    )
    synchronized(lock) {
      engine?.close()
      engine = built
      foregroundService = config.foregroundService ?: false
      config.notificationTitle?.let { notificationTitle = it }
      config.notificationText?.let { notificationText = it }
    }
    Log.i(LOG_TAG, "loaded ${specs.size} keyword(s): ${specs.joinToString { "${it.keyword}[w=${built.keywords.first { k -> k.keyword == it.keyword }.window}]" }}, vad=$vadThreshold, foregroundService=$foregroundService")
  }

  override fun unload() {
    stop()
    synchronized(lock) {
      engine?.close()
      engine = null
    }
  }

  override fun start(): Promise<Unit> = Promise.async {
    if (engine == null) throw WakeWordException("Call load() before start()")
    if (!hasMicrophonePermission()) throw WakeWordException("Microphone permission not granted")
    synchronized(lock) {
      if (capture?.isRunning == true) return@async
      val newCapture = AudioCapture(
        chunkSize = OpenWakeWordEngine.CHUNK_SIZE,
        onChunk = { chunk -> processChunk(chunk) },
        onError = { message -> errorListener?.invoke(message) },
      )
      if (foregroundService) {
        Log.i(LOG_TAG, "starting foreground service")
        WakeWordForegroundService.start(context, notificationTitle, notificationText)
      }
      newCapture.start()
      Log.i(LOG_TAG, "listening")
      capture = newCapture
    }
  }

  override fun stop() {
    synchronized(lock) {
      capture?.stop()
      capture = null
      if (foregroundService) {
        try {
          WakeWordForegroundService.stop(context)
        } catch (_: Exception) {
        }
      }
    }
  }

  override fun setThreshold(keyword: String, threshold: Double) {
    engine?.setThreshold(keyword, threshold.toFloat())
  }

  // Listeners

  override fun setDetectionListener(listener: (detection: WakeWordDetection) -> Unit) {
    detectionListener = listener
  }

  override fun clearDetectionListener() {
    detectionListener = null
  }

  override fun setScoreListener(listener: (keyword: String, score: Double) -> Unit) {
    scoreListener = listener
  }

  override fun clearScoreListener() {
    scoreListener = null
  }

  override fun setErrorListener(listener: (message: String) -> Unit) {
    errorListener = listener
  }

  override fun clearErrorListener() {
    errorListener = null
  }

  // Permissions

  override fun hasMicrophonePermission(): Boolean =
    ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED

  override fun requestMicrophonePermission(): Promise<Boolean> {
    val promise = Promise<Boolean>()
    if (hasMicrophonePermission()) {
      promise.resolve(true)
      return promise
    }
    val activity = context.currentActivity as? PermissionAwareActivity
    if (activity == null) {
      promise.reject(WakeWordException("No current activity to request permission from"))
      return promise
    }
    val listener = PermissionListener { requestCode, _, grantResults ->
      if (requestCode != PERMISSION_REQUEST_CODE) return@PermissionListener false
      promise.resolve(grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED)
      true
    }
    activity.requestPermissions(arrayOf(Manifest.permission.RECORD_AUDIO), PERMISSION_REQUEST_CODE, listener)
    return promise
  }

  // Processing (capture thread)

  private fun processChunk(chunk: FloatArray) {
    val engine = engine ?: return
    val result = engine.process(chunk)
    scoreListener?.let { listener ->
      for ((keyword, score) in result.scores) listener(keyword, score.toDouble())
    }
    detectionListener?.let { listener ->
      for (detection in result.detections) {
        listener(WakeWordDetection(detection.keyword, detection.score.toDouble(), detection.timestamp))
      }
    }
  }
}

File: ios/HybridWakeWord.swift
import AVFoundation
import Foundation
import NitroModules

class HybridWakeWord: HybridWakeWordSpec {
  private let queue = DispatchQueue(label: "com.nitrowakeword.engine", qos: .userInitiated)
  private var engine: OpenWakeWordEngine?
  private var capture: AudioCapture?
  private var manageAudioSession = true
  private var pendingChunks = 0

  private var detectionListener: ((WakeWordDetection) -> Void)?
  private var scoreListener: ((String, Double) -> Void)?
  private var errorListener: ((String) -> Void)?

  // MARK: - Properties

  public var isLoaded: Bool { engine != nil }
  public var isListening: Bool { capture?.isRunning ?? false }

  // MARK: - Lifecycle

  public func load(config: WakeWordConfig) throws -> Promise<Void> {
    return Promise.async { [self] in
      self.stop()
      guard !config.models.isEmpty else {
        throw WakeWordError("config.models must contain at least one model")
      }
      guard let melURL = ModelResolver.resolve("melspectrogram.onnx"),
            let embeddingURL = ModelResolver.resolve("embedding_model.onnx") else {
        throw WakeWordError("Base models not found. Is NitroWakeWordModels.bundle in the app?")
      }
      let vadThreshold = Float(config.vadThreshold ?? 0.3)
      let vadURL = vadThreshold > 0 ? ModelResolver.resolve("silero_vad.onnx") : nil
      if vadThreshold > 0 && vadURL == nil {
        throw WakeWordError("silero_vad.onnx not found but vadThreshold > 0")
      }

      var keywords: [(keyword: String, url: URL, threshold: Float, patience: Int)] = []
      for model in config.models {
        guard let url = ModelResolver.resolve(model.model) else {
          throw WakeWordError("Model not found: \(model.model)")
        }
        keywords.append((
          keyword: model.keyword ?? ModelResolver.defaultKeyword(for: model.model),
          url: url,
          threshold: Float(model.threshold ?? 0.5),
          patience: Int(model.patience ?? 1)
        ))
      }

      let built = try OpenWakeWordEngine(
        melURL: melURL,
        embeddingURL: embeddingURL,
        vadURL: vadURL,
        vadThreshold: vadThreshold,
        refractoryMs: config.refractoryMs ?? 1000,
        keywords: keywords
      )
      self.queue.sync {
        self.engine = built
        self.manageAudioSession = config.manageAudioSession ?? true
      }
    }
  }

  public func unload() throws {
    stop()
    queue.sync { engine = nil }
  }

  public func start() throws -> Promise<Void> {
    return Promise.async { [self] in
      guard self.engine != nil else {
        throw WakeWordError("Call load() before start()")
      }
      guard try self.hasMicrophonePermission() else {
        throw WakeWordError("Microphone permission not granted")
      }
      if self.capture?.isRunning == true { return }

      let capture = AudioCapture(
        chunkSize: OpenWakeWordEngine.chunkSize,
        manageSession: self.manageAudioSession,
        onChunk: { [weak self] chunk in self?.enqueue(chunk) },
        onError: { [weak self] message in self?.errorListener?(message) }
      )
      try capture.start()
      self.capture = capture
    }
  }

  public func stop() {
    capture?.stop()
    capture = nil
  }

  public func setThreshold(keyword: String, threshold: Double) throws {
    queue.async { [weak self] in
      self?.engine?.setThreshold(keyword: keyword, threshold: Float(threshold))
    }
  }

  // MARK: - Listeners

  public func setDetectionListener(listener: @escaping (WakeWordDetection) -> Void) throws {
    detectionListener = listener
  }

  public func clearDetectionListener() throws {
    detectionListener = nil
  }

  public func setScoreListener(listener: @escaping (String, Double) -> Void) throws {
    scoreListener = listener
  }

  public func clearScoreListener() throws {
    scoreListener = nil
  }

  public func setErrorListener(listener: @escaping (String) -> Void) throws {
    errorListener = listener
  }

  public func clearErrorListener() throws {
    errorListener = nil
  }

  // MARK: - Permissions

  public func hasMicrophonePermission() throws -> Bool {
    if #available(iOS 17.0, *) {
      return AVAudioApplication.shared.recordPermission == .granted
    }
    return AVAudioSession.sharedInstance().recordPermission == .granted
  }

  public func requestMicrophonePermission() throws -> Promise<Bool> {
    let promise = Promise<Bool>()
    if #available(iOS 17.0, *) {
      AVAudioApplication.requestRecordPermission { granted in
        promise.resolve(withResult: granted)
      }
    } else {
      AVAudioSession.sharedInstance().requestRecordPermission { granted in
        promise.resolve(withResult: granted)
      }
    }
    return promise
  }

  // MARK: - Processing

  private func enqueue(_ chunk: [Float]) {
    // Drop audio instead of building an unbounded backlog on slow devices.
    if pendingChunks > 8 { return }
    pendingChunks += 1
    queue.async { [weak self] in
      guard let self else { return }
      defer { self.pendingChunks -= 1 }
      guard let engine = self.engine else { return }
      do {
        let result = try engine.process(chunk)
        if let scoreListener = self.scoreListener {
          for (keyword, score) in result.scores {
            scoreListener(keyword, Double(score))
          }
        }
        if let detectionListener = self.detectionListener {
          for detection in result.detections {
            detectionListener(WakeWordDetection(
              keyword: detection.keyword,
              score: Double(detection.score),
              timestamp: detection.timestamp
            ))
          }
        }
      } catch {
        self.errorListener?("Inference failed: \(error.localizedDescription)")
      }
    }
  }
}
